src/Http/Routing/src/EndpointRoutingMiddleware.cs

```diff
@@ -102,6 +108,35 @@ private Task SetRoutingAndContinue(HttpContext httpContext)
             }
 
             Log.MatchSuccess(_logger, endpoint);
+
+            var shortCircuitMetadata = endpoint.Metadata.GetMetadata<ShortCircuitMetadata>();
+            if (shortCircuitMetadata is not null)
+            {
+                if (!_routeOptions.SuppressCheckForUnhandledSecurityMetadata)
+                {
+                    if (endpoint.Metadata.GetMetadata<IAuthorizeData>() is not null)
+                    {
+                        ThrowMissingAuthMiddlewareException(endpoint);
+                    }
+
+                    if (endpoint.Metadata.GetMetadata<ICorsMetadata>() is not null)
+                    {
+                        ThrowMissingCorsMiddlewareException(endpoint);
+                    }
+                }
+
+                if (shortCircuitMetadata.StatusCode.HasValue)
+                {
+                    httpContext.Response.StatusCode = shortCircuitMetadata.StatusCode.Value;
+                }
+
+                if (endpoint.RequestDelegate is not null)
+                {
+                    return endpoint.RequestDelegate(httpContext);
+                }
+
```

Comment:
There must be a more efficient way to do this since it's a first class feature now.

Reply:
Which part concerns you, `endpoint.Metadata.GetMetadata<ShortCircuitMetadata>();`?

Reply:
Thinking a bit further on this, I believe this will hurt performance in the common case (that your route matches) compared to the more exceptional case that you want to shortcircuit. I think that probably means that this should be opt-in, so that apps that don't have this concern don't have to pay the performance penalty.

An alternative design is to create and set a special type of endpoint `ShortcutEndpoint` with a fixed delegate and a `Shortcut` property, and cast to that instead of searching inside the metadata collection.

Reply:
Our goal should be to keep logic that is not related to endpoint selection outside of routing as much as possible. Code that provides specific behaviors should in general go into their own middleware so that devs get a chance of choosing when to apply the behavior and so that routing doesn't have to be aware of such concerns.

Reply:
What part is the perf overhead, `endpoint.Metadata.GetMetadata<ShortCircuitMetadata>();`? Rather than separating the components, we could mitigate that cost instead.

Reply:
@Kahbazi can you add a microbenchmark for this middleware that shows the cost of the new `GetMetadata<ShortCircuitMetadata>();` check for endpoints that don't short circuit vs those that do?

Reply:
If GetMetadata is too slow then we should look into improving it, It gets called a lot per request.

Alternatives:
- A new bool property on Endpoint?
- A derived Endpoint?

Reply:
The approach that is currently followed here is inheritance, see "Endpoint" and "RouteEndpoint".

Reply:
Inheritance doesn't seem like a good fit in this case, we'd want this behavior to apply to existing kinds of endpoints, especially RouteEndpoint's.

Reply:
The more this grows, the more it feels that the execution part should be in its own middleware, and we should keep the routing middleware as is.

It feels that this is a very specific behavior that is useful in a few cases but that everybody is now paying for. 

Why does this deserve a special treatment within the endpoint routing middleware as compared to the authorization/authentication or other similar middleware. If the concern is that there needs to be an extra call in the pipeline that the user might forget, why can't it work similar to how the authentication middleware does and throw an exception if it detects the middleware is not present.

Reply:
> If the concern is that there needs to be an extra call in the pipeline that the user might forget, why can't it work similar to how the authentication middleware does and throw an exception if it detects the middleware is not present.

That ends up being the same check as this that gets called on every request.

The microbenchmarks do not show this to be significant overhead.